packages/resolve-url-loader-filesearch/index.js.js

```diff
@@ -4,151 +4,78 @@
-        .filter(function notHidden(filename) {
-          return (filename.charAt(0) !== '.');
-        })
-        .map(function toAbsolute(filename) {
-          return path.join(basePath, filename);
-        })
-        .filter(function directoriesOnly(absolutePath) {
-          return fs.existsSync(absolutePath) && fs.statSync(absolutePath).isDirectory();
-        })
-        .filter(function notInRootPath(absolutePath) {
-          return (pathToRoot.indexOf(absolutePath) < 0);
-        })
-        .filter(testNotPackage)
-        .forEach(function enqueue(absolutePath) {
-          queue.push(absolutePath);
-        });
-    }
-
-    /**
-     * Test whether the given directory is above but not equal to any of the project root directories.
-     * @param {string} absolutePath An absolute path
-     * @returns {boolean} True where a package.json or bower.json exists, else False
-     */
-    function testWithinLimit(absolutePath) {
+const path = require('path');
+const { createJoinFunction, defaultJoinCreateIterator, defaultJoinOperation } = require('resolve-url-loader');
+
+module.exports = (options = {}) => {
+  const includeRoot = !!options.includeRoot;
+  const attempts = Math.max(0, options.attempts) || 1E+9;
+  const breakOnFile = [].concat(options.breakOnFile) || ['package.json', 'bower.json'];
+
+  const predicate = typeof options.predicate === 'function' ?
+    options.testIsRunning :
+    ((fs, absolutePath) => breakOnFile
+      .map((file) => path.resolve(absolutePath, file))
+      .every((file) => !fs.existsSync(file) || !fs.statSync(file).isFile()));
+
+  const next = typeof options.createIterator === 'function' ?
+    options.createIterator :
+    defaultJoinCreateIterator;
+
+  const createSearchingIterator = (filename, uri, bases, isAbsolute, {root, fs}) => {
+    const resolvedRoot = !!root && path.resolve(root) || process.cwd();
+    const testWithinLimit = (absolutePath) => {
       var relative = path.relative(resolvedRoot, absolutePath);
       return !!relative && (relative.slice(0, 2) !== '..');
-    }
+    };
 
-    /**
-     * Print verbose debug info where <code>options.debug</code> is in effect.
-     * @param {string} result Final text to append to the message
-     */
-    function flushMessages(result) {
-      if (log) {
-        var text = ['\n' + PACKAGE_NAME + ': ' + uri]
-          .concat(messages)
-          .concat(result)
-          .join('\n  ');
-        log(text);
+    const enqueue = (queue, excludes, basePath) =>
+      fs.readdirSync(basePath)
+        .filter((filename) => filename.charAt(0) !== '.')
+        .map((filename) => path.join(basePath, filename))
+        .filter((absolutePath) => fs.existsSync(absolutePath) && fs.statSync(absolutePath).isDirectory())
+        .filter((absolutePath) => !excludes.contains(absolutePath))
+        .filter((absolutePath) => predicate(fs, absolutePath))
+        .forEach((absolutePath) => queue.push(absolutePath));
+
+    return function* () {
```

Comment:
@runfaj @mightypork this is untested but it is the original file-search "magic" refactored as a generator.

I'm open to ideas on what we can do (with minimum effort) to make this available/useful. And if we do what level of testing is appropriate.